config: Add default_{vcpus, memory} configurations

dvoytik opened the pull request:
Introduce two new configuration settings:
 * default_vcpus - defines default vCPU number for new created PODs
 * default_memory - defines default memory size for new created PODs

 Fixes #164.

Signed-off-by: Dmitry Voytik <dmitry.voytik@huawei.com>

jodh-intel commented:
Hi @dvoytik - thanks for raising.

Travis is failing as you seem to have found a bug/limitation of our checkcommits tool - it currently requires the "Fixes #XXX" comment in column 0 (no indent :) as that is our convention (I'll raise a PR on the tool itself to allow atleast a pure whitespace indent...)

More fundamentally, semaphore is unable to launch a hypervisor with this PR. I'm not sure if you've used it, but you can actually `ssh` into the semaphore build env for this PR to debug why that is happening.

dvoytik commented:
@jodh-intel thanks for clarification this. I'll try to fix all problems and update the PR.

devimc commented:
lgtm

jodh-intel commented:
Hi @dvoytik - the code looks good, but since you are changing virtcontainers code as well, the way we will need to land this is:

1) You raise a PR on https://github.com/containers/virtcontainers (for the changes you've made to files below `vendor/*` on this PR).
2) Once that PR has been merged into virtcontainers, we'll re-vendor the runtime to update the version of virtcontainers it uses.
3) We can then land the changes on this PR that make use of `Default{VCPUs,MemSz}` from the version of virtcontainers imported into the runtime.

dvoytik commented:
Hi @jodh-intel. Got it. I'll do it.

jodh-intel commented:
Follow-up to https://github.com/clearcontainers/runtime/pull/297#issuecomment-312874147: I've raised the following PR on `checkcommits` to resolve the issue you got caught by where you had indented the "Fixes XXX" comment:

- https://github.com/clearcontainers/tests/pull/114/
  (specifically https://github.com/clearcontainers/tests/pull/114/commits/f4a502ebea2380e26b5a2dda0d3181a0a0576a1d)

dvoytik commented:
Hi @jodh-intel, could you please merge https://github.com/containers/virtcontainers/pull/299, so I can proceed. Thank you.

dvoytik commented:
The PR was rebased and updated: changed default memory size from 512MiB to 2048MiB as @dlespiau suggested.
This PR still needs merging of https://github.com/containers/virtcontainers/pull/299 and vendoring of virtcontainers.
Something is wrong with both CI. Please let me know if I should do something about it. Thanks.

jodh-intel commented:
Adding tag to ensure we don't inadvertently merge this before https://github.com/containers/virtcontainers/pull/299 lands...

dvoytik commented:
I was thinking, how to let the sysadmin/ops easily configure the runtime to have by default as many vCPUs as physical cores on the host. What about making **_default_vcpus_** as signed integer and have special meaning for -1 - automatically inferring from physical core count. Also since QEMU supoorts no more than 255 cores, we should also limit this setting.

jodh-intel commented:
@dvoytik - that sounds like a good idea.

dvoytik commented:
@jodh-intel, Ok then. I'll update this PR after https://github.com/containers/virtcontainers/pull/299 is being merged and vendored into runtime repo.

dvoytik commented:
Update:
 - added vendoring of virtcontainers
 - now **_default_vcpus_** can be set to -1 to enable automatic physical cores infer
P.S.
It seems I wasn't the reason of CI failure this time :smile:

grahamwhaley commented:
Hi @dvoytik Indeed, that was not you breaking the CI - this was a problem we seemed to be seeing last week with Travis. I've not seen one yet this week, so it could have been transient. Something that often fixes it seems to be a rebase and push, or a minor change and push (to get a SHA change, to get a rebuild of the CI). See https://github.com/clearcontainers/tests/pull/136 for a few more details, but we've not bottomed the issue yet.

dvoytik commented:
Hi @grahamwhaley. OK, thanks for clarifying this. Just please let me know when I should repush this PR to trigger CI tests again, I don't want to spam this PR until. Thank you in advance.

dvoytik commented:
A clean re-base of the PR. It seems that now all CIs passed.

grahamwhaley commented:
Hi @dvoytik Just to note, we are not ignoring you ;-) - I think this looks good now. Just waiting for @jodh-intel to come back online, do the review, and drop the DNM tag. That should be early next week I believe.
I'll note that https://github.com/containers/virtcontainers/pull/299 has landed, so I believe we are all clear to get this moving again.

dvoytik commented:
@grahamwhaley, thanks! :)

devimc commented:
lgtm

devimc commented:
@jodh-intel could you review it and remove the ```do-not-merge``` tag if you accept this pr?

jodh-intel commented on Makefile:
```diff
@@ -92,6 +92,9 @@ PAUSEBINRELPATH := bin/pause
 
 GLOBALLOGPATH := $(PKGLIBDIR)/runtime/runtime.log
 
+DEFDEFVCPUS := 1
```
I'm not clear on the "double DEF" here. Couldn't this just be `DEFAULTVCPUS` and `DEFAULTMEMSZ`?

dvoytik replied:
It's a default of default value :) If you insist, a can change it, no problem

jodh-intel commented on Makefile:
```diff
@@ -92,6 +92,9 @@ PAUSEBINRELPATH := bin/pause
 GLOBALLOGPATH := $(PKGLIBDIR)/runtime/runtime.log
 
+DEFDEFVCPUS := 1
+DEFDEFMEMSZ := 2048
```
I'd add a comment to specify the units this variable is using.

dvoytik replied:
OK.

jodh-intel commented on Makefile:
```diff
@@ -178,6 +183,9 @@ const defaultRuntimeRun = "$(PKGRUNDIR)"
 const defaultShimPath = "$(SHIMPATH)"
 const pauseBinRelativePath = "$(PAUSEBINRELPATH)"
 
+const defaultDefaultVCPUs uint32 = $(DEFDEFVCPUS)
```
Again, `defaultDefault...` looks a little odd to me. Maybe just `defaultVCPUs`?

dvoytik replied:
There is already a function called defaultVCPUs(). Wouldn't it be confusing having both with the same name? If not, then I can change as you propose.

jodh-intel replied:
Agreed. You could use a name like `const defaultVCPUCount` maybe.

dvoytik replied:
Ok. Will do it.

jodh-intel commented:
Thanks @dvoytik.

lgtm.

jodh-intel commented:
hmm, pullapprove is confused again. Let's try the label toggle again...

jodh-intel commented:
Yep - convinced it.
